### PR TITLE
`nix derivation add`, `show-derivation` -> `derivation show`

### PR DESCRIPTION
**Pull request by Ericson2314**
# Motivation

Implementing approved #7828

TODO:

 - [x] Fix `DerivationTest.Derivation_impure_from_json` unit test test
 - [x] Need to validate derivation so we don't write illegal ones to the store
 - [x] Write documentation (it is just copied right now)

# Context

#7828

Grouping these commands was also discussed in #7903

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [x] agreed on idea
 - [x] agreed on implementation strategy
 - [x] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [x] documentation in the manual
 - [x] code and comments are self-explanatory
 - [x] commit message explains why the change was made
 - [x] new feature or incompatible change: updated release notes

### CODE REVIEW

**Comment by roberth** [comment]
I believe this command should be under `store`, like `add-file`. While that wouldn't be consistent with `nix show-derivation`, I believe that this command should also be moved. By doing so, the top level namespace is relatively clean, which helps with discoverability. It also follows the noun-verb naming rule a bit better.

**Comment by Ericson2314** [comment]
I opened #7903 for the naming of this command.

**Comment by Ericson2314** [comment]
OK the check boxes are checked! This still sucks in many ways

 - Derivation name is specified via CLI flag
 - Input-addressed output paths must be manually computed
 - Documentation sucks --- we should document the derivation format separately from these utility command man pages

But it's an experimental command, so I am tempted to say we should just land a beachhead and then improve it.

File: src/nix/add-derivation.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+R""(
+
+# Description
+
+This command reads from the given path a JSON representation of a
```
Comment (Ericson2314) [inline src/nix/add-derivation.md:5]:
```suggestion
This command reads from standard input a JSON representation of a
```

**Comment by fricklerhandwerk** [comment]
Triaged in the Nix team meeting 2023-03-23:

- @edolstra: Does this import only one derivation at a time?
- @ericson2314: Production quality / High performance language interface would use bindings
    - this is for easy experimentation and making the CLI complete on principle
    - It's a limitation, but OK for the moment
- idea approved, assigned to @roberth

**Review by roberth — CHANGES_REQUESTED** [review]
We need to resolve the `--name` situation first.
Adding a prototype of this command first would be counterproductive, because we know that it barely serves a purpose. It only serves to annoy the poor soul who tries to use it before it's done. (and the test case I suppose, although like 80% of the test value is already delivered: showing that the code is mostly ok but the design isn't right)

File: src/libstore/derivations.cc [hunk 2]
```diff
@@ -950,4 +1078,41 @@ nlohmann::json Derivation::toJSON(const Store & store) const
     return res;
 }
 
+
+Derivation Derivation::fromJSON(
+    const Store & store, std::string_view drvName,
```
[the remark this answers is not in the document]
Comment (Ericson2314) [reply src/libstore/derivations.cc]: There is now a `name` field!

File: src/nix/add-derivation.md (hunk 1 above)
Comment (roberth) [reply src/nix/add-derivation.md:5]: It's back :ghost:

**Review by roberth — COMMENTED** [review]
Looks alright.
TODO
 - [ ] Apply the suggestion in add-derivation.md

Scoped out, prio mine.

 - should have: Rename the command
 - nice to have: Streaming JSON Lines

File: src/nix/add-derivation.md (hunk 1 above)
Comment (Ericson2314) [reply src/nix/add-derivation.md]: And it's gone again! :ghost:

**Comment by fricklerhandwerk** [comment]
Isn't this a plumbing command? `nix store add-derivation`. @Ericson2314 last time you proposed having only convenience wrappers at the top level.

**Comment by Ericson2314** [comment]
@fricklerhandwerk Since we already have `nix show-derivation` I soon as well add this first and then separately go about finding a new home for both of them.

In generally I rather separate the tasks of making sure we have commands for missing functionality, and reorganizing the commands, as the former is somewhat easier and the latter requires more team thinking which can keep open PRs for an indefinite amount of time if it blocks the former.

**Comment by roberth** [comment]
Doing it this way does add more tech debt in the form of either missing release notes or potentially giving users a command the we're going to break.
Not sure if refinancing the tech debt is progress.
Ideally we can rename it now, but otherwise I lean slightly towards merging? Stagnation causes even more suffering.

**Comment by fricklerhandwerk** [comment]
Side note: Progress is not a value in itself. No change can also mean stability one can build on. Otherwise I agree on the tech dept, let's keep it minimal.

**Comment by Ericson2314** [comment]
If you think it's fine I'll go ahead and do `nix derivation {add,show}` right away.

**Comment by Ericson2314** [comment]
I went ahead and did that.

Do note that I think there are still other breaking changes that would happen, relating to the fact that the JSON format for derivations (used by `show-dervation`) doesn't yet follow the API guidelines.

**Review by roberth — APPROVED** [review]

**Pull request merged**